### ISSUE TITLE
Cannot rebind any gamepad controls - 0.9.1-3

### ISSUE THREAD

**Issue by aaronfranke** [supertuxkart/stk-code#2617]
I'm using a Steam controller set to output Xbox 360 controller inputs and SuperTuxKart 0.9.1-3 from the repos on Xubuntu 16.04 LTS. 

I can navigate the menus fine, but when I try to re-bind a key, I select it, then press A, then a dialog box flashes for a frame or two, and then disappears, leaving me unable to re-bind the action. The same thing happens if I use a keyboard or mouse to select it (with enter button or with mouse click). I just installed SuperTuxKart.

**Comment by qwertychouskie** [comment]
Try the latest version from this PPA: https://launchpad.net/~stk/+archive/ubuntu/dev

**Comment by MTres19** [comment]
Duplicate of #2557 and #2240. However, the title to this issue may be more relevant. I don't think it's specific to a single controller.

**Comment by MTres19** [comment]
However, could someone check if this affects keyboards too or only gamepads? Thanks.

**Comment by aaronfranke** [comment]
Ok, so I have some more information to add. This happens on my system for ALL gamepads, but not the keyboard. Tested gamepads (probably irrelevant, but I'll post anyway): Generic Thrustmaster Firestorm, PS2 controller with USB adapter, GameCube controller with USB adapter, generic Android controller.

@qwertychouskie I tried doing as you said and installed the version from that PPA, "0.9.2-1build2~ubuntu16.04.1", and can confirm that this bug happens there. I can rebind the keyboard, however, I cannot rebind any of my gamepads at all.

**Comment by MTres19** [comment]
It seems to be a Linux-only problem. I ran STK under valgrind; I'll post the output soon.

**Comment by MTres19** [comment]
Here's the output from valgrind.

[stk-valgrind-joystick-config.txt](https://github.com/supertuxkart/stk-code/files/452731/stk-valgrind-joystick-config.txt)

**Comment by MTres19** [comment]
@deveee Any ideas?

**Comment by MTres19** [comment]
Here's another valgrind log, this time with `--track-origins=yes`.
[stk-valgrind-joystic-config-track-origins.txt](https://github.com/supertuxkart/stk-code/files/491217/stk-valgrind-joystic-config-track-origins.txt)

**Comment by auriamg** [comment]
Valgrind would only help if this was a memory bug (crash, etc.)
What is much more likely is that STK is getting "erratic" readings (or just unexpected readings) from the gamepad or the driver and has trouble figuring out what button you pressed (or didn't). So running valgrind is never a bad idea but it's far from certain that this kind of bug can be debugged with valgrind at all, it would be more useful for someone who can reproduce this issue to step through the input sensing code and see what happens

**Comment by MTres19** [comment]
I'd better learn how to use GDB then.

**Comment by MTres19** [comment]
What is strange is that when I copied the input configuration from Windows
I never experienced any gameplay issues. I will attempt to use GDB though
if I have time.

**Comment by Danfun64** [comment]
For me the same issues exist in Windows as well as Linux.

**Comment by hiker** [comment]
I was able to reproduce it last night (while doing something else, so didn't chase it up). As mentioned in #2557 it seems to be easier to reproduce when more than one gamepad is connected.

**Comment by hiker** [comment]
@auriamg I think I understand the problem now: 
the input manager uses two arrays to store if the previous values of an axis was high (m_sensed_input_high_gamepad), using the axis id (called button in inputSensing()).
Now when you have two gamepads connected, what can happen when you try to define a setting for the second controller:
the second controller marks an axis 'A' as high. In the next frame, the first controller reports the axis A to be low. This change (from high to not high) triggers the call to OptionsScreenDevice::gotSensedInput() for the first controller. But this code detects that the controller that delivered the event has a different name than the controller you are trying to configure (assuming you have two different controllers connected), so the new setting is ignored, and the dialog is closed. The following patch seems to fix that issue:
```
git diff src/input/input_manager.cpp
diff --git a/src/input/input_manager.cpp b/src/input/input_manager.cpp
index f41aafab6..61f025262 100644
--- a/src/input/input_manager.cpp
+++ b/src/input/input_manager.cpp
@@ -489,6 +489,7 @@ void InputManager::inputSensing(Input::InputType type, int deviceID,
         // This is done by storing it as a sign (and since button can
         // be zero, we add one before changing the sign).
         int input_id = value>=0 ? 1+button : -(1+button);
+        input_id += 200 * deviceID;

         bool id_was_high         = m_sensed_input_high_gamepad.find(input_id)
                                    != m_sensed_input_high_gamepad.end();
```
The idea is that each controller axis gets a unique number (axis of controller + deviceId*200), so two controllers can not get mixed up anymore. I picked 200 since the HAT axis are reported as axis 100 and 101, so the value must be larger than 101.
@auriamg can you check if this is the best fix?

**Comment by hiker** [comment]
Note that this can also happen if there should be a device reporting an axis (harddrivers, or in my case an invalid gamepad axis that is constantly reported as -32767).

**Comment by auriamg** [comment]
@hiker Could you try branch Fix-2617 ? That should fix the issue in a cleaner way, though I was not able to fully test

**Comment by hiker** [comment]
Fixed in 626f540b1516376e997389369d16fd9a3cad4bbb.
